iroh-net/src/relay/server/metrics.rs

```diff
@@ -117,3 +133,161 @@ impl Metric for Metrics {
+                if let Some(client) = r {
+                    let mut clients = writer_clients.lock().await;
+                    clients.insert(client);
+                }
+            }
+        });
+
+        let batch = Arc::clone(&clients);
+        let rclient_clone = rclient.clone();
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(std::time::Duration::from_secs(30));
+            let relay_key = format!("unique_nodes_{}", redis_suffix.clone());
+            let global_key = "unique_nodes".to_string();
+
+            async fn batch_update_redis(
+                client: &redis::Client,
+                batch: &HashSet<PublicKey>,
+                relay_key: &str,
+                global_key: &str,
+            ) -> redis::RedisResult<()> {
+                let mut conn = client.get_multiplexed_async_connection().await?;
+                let mut pipeline = redis::pipe();
+                for node_id in batch.iter() {
+                    let node_id_str = hex::encode(node_id.as_bytes());
+                    pipeline
+                        .sadd(format!("{}_1d", relay_key), node_id_str.clone())
+                        .ignore();
+                    pipeline
+                        .sadd(format!("{}_1d", global_key), node_id_str.clone())
+                        .ignore();
+                    pipeline
+                        .sadd(format!("{}_7d", global_key), node_id_str.clone())
+                        .ignore();
+                    pipeline
+                        .sadd(format!("{}_30d", global_key), node_id_str.clone())
+                        .ignore();
+                }
+                pipeline.expire(format!("{}_1d", relay_key), 86400).ignore(); // 1 day
+                pipeline
+                    .expire(format!("{}_1d", global_key), 86400)
+                    .ignore(); // 1 day
+                pipeline
+                    .expire(format!("{}_7d", global_key), 604800)
+                    .ignore(); // 7 days
+                pipeline
+                    .expire(format!("{}_30d", global_key), 2592000)
+                    .ignore(); // 30 days
+                pipeline.query_async(&mut conn).await
+            }
+
+            async fn get_unique_nodes(
+                client: &redis::Client,
+                key: &str,
+            ) -> redis::RedisResult<u64> {
+                let mut conn = client.get_multiplexed_async_connection().await?;
+                let count = conn.scard(key).await?;
+                Ok(count)
+            }
+
+            loop {
+                interval.tick().await;
+                let batch = Arc::clone(&batch);
+                let rclient_clone = rclient_clone.clone();
+                let relay_key = relay_key.clone();
+                let global_key = global_key.clone();
+                tokio::time::timeout(Duration::from_secs(60), async move {
+                    {
+                        let mut batch = batch.lock().await;
```

Comment:
I'm not sure why you go for this complicated architecture of having a local hashset which you then fully iterate over every time to send to the relay server.  Most of these will be duplicates since iroh-net nodes tend to be connected to a relay server for a long time.  And the way this takes the lock ends up with the messages being buffered in a channel while the lock is held here (which is being held even while a connection to the redis server is being checked out - way to long a critical section).


Anyway, what is the benefit of batching?  I suspect this is counter-benficial.

I'd suggest an architecture where there's an actor that runs for the lifetime of the magicsock.  The actor receives messages with NodeIds.  The actor also maintains a connection to the redis server and each time it receives an update it sends it to the redis server.  It should do this in a way so that the sender of the messages is never blocked.

I think the logic flow will be a lot easier to follow, all  state is on the redis server so less memory needed locally.